rustls-mbedcrypto-provider/src/fips_utils.rs

```diff
@@ -0,0 +1,256 @@
+    let mut rng = crate::rng::rng_new().ok_or(rustls::Error::FailedToGetRandomBytes)?;
+    fips_check_ec_pub_key_mbed(ec_mbed_pk, &mut rng).map_err(wrap_mbedtls_error_as_fips)?;
+    log::info!("ECC Full Public-Key Validation: passed");
+    Ok(())
+}
+
+/// ECC Pairwise Consistency Test upon generation that mimics the shared
+/// secret computation using the recent generated key pair and a known
+/// key pair with the same domain parameters and comparing the shared
+/// secret computation values calculated using either FFC DH or ECC CDH
+/// primitives.
+///
+/// Learn more in:
+/// - [FIPS 140-3 IG] section 10.3.A.
+/// - [SP 800-56Ar3] section 5.6.2.1.4.
+///
+/// [FIPS 140-3 IG]: https://csrc.nist.gov/projects/cryptographic-module-validation-program/fips-140-3-ig-announcements
+/// [SP 800-56Ar3]: https://csrc.nist.gov/pubs/sp/800/56/a/r3/final
+pub(crate) fn fips_ec_pct(ec_mbed_pk: &mut Pk, ec_group_id: EcGroupId) -> Result<(), rustls::Error> {
+    let mut rng = crate::rng::rng_new().ok_or(rustls::Error::FailedToGetRandomBytes)?;
+    // Get a static ec pub key based on given [`EcGroupId`]
+    let known_ec_key = get_known_ec_key(&ec_group_id).expect("validated");
+    let mut known_ec_key = known_ec_key
+        .lock()
+        .map_err(|_| rustls::Error::General("Failed to get known ec key: poisoned lock".to_string()))?;
+
+    fips_ec_pct_mbed(ec_mbed_pk, &mut known_ec_key, &mut rng).map_err(wrap_mbedtls_error_as_fips)?;
+    log::info!("ECC Pairwise Consistency Test: passed");
+    Ok(())
+}
+
+#[allow(non_snake_case)]
+fn fips_ec_pct_mbed<F: mbedtls::rng::Random>(
+    ec_mbed_pk: &mut Pk,
+    known_ec_key: &mut Pk,
+    rng: &mut F,
+) -> Result<(), mbedtls::Error> {
+    // Pairwise Consistency Test upon generation that mimics the
+    // shared secret computation using the recent generated key
+    // pair and a known key pair with the same domain parameters
+    // and comparing the shared secret computation values
+    // calculated using either FFC DH or ECC CDH primitives.
+    let mut shared_1 = vec![0; ECDSA_MAX_LEN];
+    let mut shared_2 = vec![0; ECDSA_MAX_LEN];
+    let len = ec_mbed_pk.agree(known_ec_key, &mut shared_1, rng)?;
+    shared_1.truncate(len);
+    let len = known_ec_key.agree(ec_mbed_pk, &mut shared_2, rng)?;
+    shared_2.truncate(len);
+    if shared_1 != shared_2 {
+        return Err(mbedtls::Error::EcpInvalidKey);
+    }
+    // For an ECC key pair (d, Q): Use the private key, d, along
+    // with the generator G and other domain parameters
+    // associated with the key pair, to compute dG (according to
+    // the rules of elliptic-curve arithmetic). Compare the
+    // result to the public key, Q. If dG is not equal to Q,
+    // then the pair-wise consistency test fails.
+    let mut ec_group = ec_mbed_pk.ec_group()?;
+    let d = ec_mbed_pk.ec_private()?;
+    let Q = ec_mbed_pk.ec_public()?;
+    let G = ec_group.generator()?;
+    let dummy_ec_point = EcPoint::new()?;
```

Comment:
Why this `dummy_ec_point`? I think it is the infinity point, so you're setting `dG` to the identity if `mul_with_rng` fails.

Reply:
So it should be fine to return error directly?

Reply:
fixed